doc: Added documentation and diagrams for HostedCluster and NodePool (#1863)
openshift/hypershift: docs/content/how-to/aws/index.md

        - 1 Public Subnet (for Public HostedCluster)
        - 1 Internet Gateway
        - 1 NAT Gateway
        - 1 Security Group (Worker Nodes)
        - 1 Route Tables (Public)
        - 1 Public Hosted Zone for Cluster Ingress

    === "Private"

        - 1 VPC
        - 1 DHCP Options
        - 1 Private Subnet (for Private HostedCluster)
        - 1 Internet Gateway
        - 1 NAT Gateway
        - 1 Security Group (Worker Nodes)
        - 1 Route Tables (Private)
        - 1 Private Hosted Zone for PrivateLink

    === "PublicAndPrivate"
        - 1 VPC
        - 1 DHCP Options
        - 2 Subnets
            - 1 Private Subnet (for Private HostedCluster)
            - 1 Public Subnet (for Public HostedCluster)
        - 1 Internet Gateway
        - 1 NAT Gateway
        - 1 Security Group (Worker Nodes)
        - 2 Route Tables (1 Private, 1 Public)
        - 2 Hosted Zones
            - 1 for Cluster Ingress (for Public clusters)
            - 1 for PrivateLink (for private clusters)

After the creation of these AWS Objects and to perform the relationship among them, you can create the Hosted Cluster CR in order to begin the Control Plane deployment.

> An easy way to see how the relationships are made among the AWS Objects you can check this CLI HostedCluster deployment

Author: Let me rephrase it:

```
You can check also an alternative and automatic way for a HostedCluster provisioning in AWS provider using the CLI (maybe it's also an useful way to have an initial view of the objects needed by the HostedCluster to be properly provisioned in the AWS side).
```

<Timestamp>  INFO  Created role	{"name": "jparrill-dest-zjhcn-worker-role"}
<Timestamp>  INFO  Created instance profile	{"name": "jparrill-dest-zjhcn-worker"}
<Timestamp>  INFO  Added role to instance profile	{"role": "jparrill-dest-zjhcn-worker-role", "profile": "jparrill-dest-zjhcn-worker"}
<Timestamp>  INFO  Created role policy	{"name": "jparrill-dest-zjhcn-worker-policy"}
<Timestamp>  INFO  Created IAM profile	{"name": "jparrill-dest-zjhcn-worker", "region": "us-west-1"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "Namespace", "namespace": "", "name": "jparrill"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "Secret", "namespace": "jparrill", "name": "jparrill-dest-pull-secret"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "", "namespace": "jparrill", "name": "jparrill-dest"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "Secret", "namespace": "jparrill", "name": "jparrill-dest-etcd-encryption-key"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "Secret", "namespace": "jparrill", "name": "jparrill-dest-ssh-key"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "NodePool", "namespace": "jparrill", "name": "jparrill-dest-us-west-1b"}
```

</details>

> After this, the Openshift Hosted Control Plane will continue being deployed and eventually the cluster will create some resources which will end on AWS object creation (These are not managed by Hypershift but Openshift):

Author: yep, I was talking about the ELB created by the HostedControlPlane as an Ingress ELB. Let me rephrase it to make it clear:

"After this, the Openshift Hosted Control Plane will continue being deployed and eventually it will create some other resources which will end on AWS object creation (ELB) (These are not managed by Hypershift but Openshift):"

WDYT?

> - **Cluster Route**: It's the main router for the accesses from outside in Openshift. **It creates 1 ELB (External Load Balancer) in AWS** provider which route the request to the cluster from outside.

Author: AFAIK The IngressController, I'm I right?

- **Operator**: Stands for Hypershift Operator and it has some controllers inside which manages the HostedClusters and NodePools created in the cluster's namespace.

!!! info
    Regardless of whether you use the External DNS feature or not, as mentioned earlier, you need an already registered (or at least delegated) DNS domain in *Route 53*. This is where the public records are created to allow external access

### HostedCluster

This is the Hosted Cluster namespace where you as a user will create the definition of your cluster and your nodes. An easy way to create a hosted cluster is by using the CLI as we've seen in the last sample.

HostedCluster and NodePool controllers (part of Hypershift Operator) are monitoring the HostedCluster and NodePool objects being created in the whole cluster. So once you create one of these components, the controllers will begin the reconciliation.

You will need to have in mind that some more objects like Secrets and ConfigMaps (described in the diagram) are needed for a successful `HostedCluster` deployment.

> From these mentioned objects, the Hypershift operator will create some more in 2 places, Openshift infrastructure and also in AWS Cloud Provider. Let's take a look to the following diagram to discover which objects are created in Openshift.

Reviewer: In OpenShift or by OpenShift?

                CA[Certificates]
                SCETC[...]
            end

            subgraph PVC[PVCs]
                etcd
                ovn
            end

            subgraph APPS[ControlPlane Apps]
                STS[Stateful Sets]
                DP[Deployments]
                RS[Replica Sets]
            end

            HCP <-.-> CL <-.-> AWSCL
            subgraph HCP[Hosted Control Plane]
                DescHCP("Contains the OCP
                HostedCluster details")
            end
            subgraph AWSCL[AWSCluster]
                DescAWSC("Contains the AWS details like
                VPC, Region, Subnet, etc...")
            end
            subgraph CL[Cluster]
                DescCL("Perform the relationship between
                AWSCluster and HostedControlPlane")
            end
        end
    end
```
!!! note "Legend"
    - **Orange:** OCP objects directly related with AWS.
    - **Cyan:** Provider agnostic OCP objects.
    - The *dotted* lines are associations among AWS Components created by the CLI or the Operator
    - The **thick** lines are objects created by the CLI into AWS platform and needed by the Hypershift Operator to work properly

> At the same time, the Hypershift operator will expect to exists the mentioned Infra components in the defined provider (You could use the Hypershift CLI in order to create them automatically). As a first step it will assign an InfraID to all the objects created in the AWS side.

Reviewer: Are all the things in the orange box created by the hypershift operator?

> - **Diagram:** The Diagrams shows 2 things, the objects Needed and Created in the AWS infrastructure. After this one we will dissect the objects in a more detailed way.

Reviewer 2: I'm confused about the ingress controller in that picture. FWIW The ingress controller runs in the hosted cluster.

Author: Totally right. My bad

- **Diagram:** The diagram shows the NodePool resultant objects created into the AWS Infrastructure.

```mermaid
flowchart LR
    style HO fill:#BAE0E9,stroke:#333,stroke-width:3px
    style EC2 fill:#FEA175,stroke:#333,stroke-width:3px
    style IAM fill:#FEA175,stroke:#333,stroke-width:3px
    subgraph OCP["Openshift Management Cluster"]
        direction LR
        HO[fa:fa-server Hypershift Operator]
    end
    HO ==> EC2
    HO ==> IAM
    EC2I1 -.-> IAMR
    EC2I2 -.-> IAMR
    EC2I1 -.-> IAMP
    EC2I2 -.-> IAMP
    subgraph ACP["AWS Cloud Provider"]
        direction LR
        subgraph EC2
            EC2I1[EC2 Instance]
            EC2I2[EC2 Instance]
        end
        subgraph IAM
            IAMR[Role Policy]
            IAMP[Profile]
        end
    end

```

!!! note "Legend"
    - **Orange:** OCP objects directly related with AWS.
    - **Cyan:** Provider agnostic OCP objects.
    - The *dotted* lines are associations among AWS Components created by the CLI or the Operator
    - The **thick** lines are objects created by the CLI into AWS platform and needed by the Hypershift Operator to work properly

> ## AWS Components and Hypershift

Reviewer 2: This is the part that I'm interested in. Can we please put this in its own section within reference/provider-infrastructure

- Infra managed by hypershift in Hosted Cluster AWS account
- Infra managed by kubernetes in Hosted Cluster AWS account

#### Pre-required and Unmanaged Infra in Hosted Cluster AWS account

=== "Public"
    - 1 VPC
    - 1 DHCP Options
    - 1 Public Subnet (for Public HostedCluster)
    - 1 Internet Gateway
    - 1 NAT Gateway
    - 1 Security Group (Worker Nodes)
    - 1 Route Tables (Public)
    - 1 Ingress Service Load Balancer

=== "Private"
    - 1 VPC
    - 1 DHCP Options
    - 1 Private Subnet (for Private HostedCluster)
    - 1 Internet Gateway
    - 1 NAT Gateway
    - 1 Security Group (Worker Nodes)
    - 1 Route Tables (Private)
    - 1 Private Link

=== "PublicAndPrivate"
    - 1 VPC
    - 1 DHCP Options
    - 2 Subnets
        - 1 Private Subnet (for Private HostedCluster)
        - 1 Public Subnet (for Public HostedCluster)
    - 1 Internet Gateway
    - 1 NAT Gateway
    - 1 Security Group (Worker Nodes)
    - 2 Route Tables (1 Private, 1 Public)
    - 2 Hosted Zones
        - 1 Ingress Service Load Balancer (for Public Hosted Clusters)
        - 1 Private Link (for Private Hosted Clusters)

#### AWS Infra Managed by Hypershift

> === "Management Cluster"

Reviewer 2: can we be specific here "Management AWS account"?

>     - Kube API Server Load Balancer

Reviewer 2: As below can we present this as "aws resource type - use case", e.g "ELB - used for Kube API Server Load Balancer"

> === "Hosted Cluster"

Reviewer 2: Hosted Cluster AWS account

    - Kube API Server Load Balancer
    - For NodePools:
        - EC2 Instances
        - IAM Role Policy link (with EC2 Instances)
        - IAM Profile link (with EC2 Instances)

#### AWS Infra Managed by Kubernetes

=== "Hosted Cluster"

>     - Elastic Load Balancer

Reviewer 2: can we clarify i.e - for default ingress
Also this is missing s3 bucket for registry I believe

---
title: AWS Architecture and Resources
---

# AWS Architecture and Resources

In this section we will talk about the AWS Resources and Openshift objects generated by Hypershift and how the Controllers interact with them during a Hosted Cluster deployment.

## Prerequisites and Assumptions

> Here we will set the legend about the terms used in this document:

Reviewer 2: Please don't add another source of truth, let's refer to https://deploy-preview-1863--hypershift-docs.netlify.app/reference/concepts-and-personas/ and add what's missing there.

- **The Management Cluster**: It's the cluster which will host the Hosted Control Planes and the operator itself.

> - **Guest Clusters**: Those are the Hosted Control Plane deployed inside of the Management cluster.

Reviewer: Can we align with terminology with use -> Hosted Cluster.

- **Node Pools**: Set of nodes (0..N) which belongs to one Hosted Cluster. Once we have at least 1 worker inside of this NodePool we will have at least 1 EC2 Instance mapped in AWS. but the relationship goes in reverse, the NodePool does not knows about the Nodes or Workers. On the other hand we have the *Machines* which has a Label where the *NodePool* and *MachineSet* is configured.

In the Prerequisites side we will have 2 states:

- Prerequisites to deploy and work with Hypershift Operator
- Prerequisites to deploy a HostedCluster

### Prerequisites to deploy and work with Hypershift Operator

> - **OCP Cluster Admin account**: Regarding roles, this is required in order to allow the CLI to generate the proper SA and ClusterRoles.

Reviewer: Which persona?

> - **AWS Route 53 Domain**: This is needed in order to create reachable API via DNS name instead of using the ELB name.

Reviewer: Any need to mention public/private here?

> - **S3 Bucket**: Publicly accesible, this is mandatory in order to upload the OIDC data for the given HostedCluster.

Reviewer: Does private S3 bucket work as long as I have connectivity to it within my org?

> - **OCP Pull Secret**: This is not an AWS Object but will be mandatory to deploy the OCP CLuster

Reviewer: What it is not is clear, can you clarify or reference what "is" it?

> - **AWS Credentials**: To create resources into AWS Provider

Reviewer: What kind of credentials (ignore if you describe it later, I am reviewing linearly :)

> - **AWS Quota**: You will need enough quota to create all the AWS Objects managed by Hypershift and others managed by Openshift itself. We will provide a detailed list of resources.

Reviewer: What is enough quota, any numbers we can provide or links to where numbers are?

- 1 VPC
- 1 DHCP Options
- 1 Private Subnet
- 1 Public Subnet
- 1 Internet Gateway
- 1 NAT Gateway
- 1 Security Group (Worker Nodes)
- 2 Route Tables (1 Private, 1 Public)
- 2 Private Hosted Zones
    - 1 for Cluster Ingress (for Public clusters)
    - 1 for PrivateLink (for private clusters)

After the creation of these AWS Objects and to perform the relationship among them, you can create the Hosted Cluster CR in order to begin the Control Plane deployment.

An easy way to see how the relationships are made among the AWS Objects you can check this CLI HostedCluster deployment

<details>
<summary>Hosted Cluster deployment</summary>

- Command
```bash
./bin/hypershift create cluster aws \
    --aws-creds ~/.aws/credentials \
    --instance-type m6i.xlarge \
    --region us-west-1 \
    --auto-repair --generate-ssh \
    --name jparrill-dest \
    --namespace jparrill \
    --base-domain jpdv.aws.kerbeross.com \
    --node-pool-replicas 2 \
    --pull-secret pull_secret.json \
    --release-image quay.io/openshift-release-dev/ocp-release:4.12.0-ec.3-x86_64
```

> - Output

Reviewer: ```suggestion
- Example Output
```

<Timestamp>  INFO  Created role	{"name": "jparrill-dest-zjhcn-worker-role"}
<Timestamp>  INFO  Created instance profile	{"name": "jparrill-dest-zjhcn-worker"}
<Timestamp>  INFO  Added role to instance profile	{"role": "jparrill-dest-zjhcn-worker-role", "profile": "jparrill-dest-zjhcn-worker"}
<Timestamp>  INFO  Created role policy	{"name": "jparrill-dest-zjhcn-worker-policy"}
<Timestamp>  INFO  Created IAM profile	{"name": "jparrill-dest-zjhcn-worker", "region": "us-west-1"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "Namespace", "namespace": "", "name": "jparrill"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "Secret", "namespace": "jparrill", "name": "jparrill-dest-pull-secret"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "", "namespace": "jparrill", "name": "jparrill-dest"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "Secret", "namespace": "jparrill", "name": "jparrill-dest-etcd-encryption-key"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "Secret", "namespace": "jparrill", "name": "jparrill-dest-ssh-key"}
<Timestamp>  INFO  Applied Kube resource	{"kind": "NodePool", "namespace": "jparrill", "name": "jparrill-dest-us-west-1b"}
```

</details>

> After this, the Openshift Hosted Control Plane will continue their way in deployment terms and eventually the cluster will create some resources which will end on AWS object creation (These are not managed by Hypershift but Openshift):

Reviewer: ```suggestion
After this, the Openshift Hosted Control Plane will continue being deployed and eventually the cluster will create some resources which will end on AWS object creation (These are not managed by Hypershift but Openshift):
```

> - **Cluster Ingress**: It's the main router for the accesses from outside in Openshift. **It creates 1 ELB (External Load Balancer) in AWS** provider which route the request to the cluster from outside.

Reviewer: Is this an "Ingress" object or a "Route" object?

## Architecture and Workflow

> In order to deploy a HostedCluster you have two options, using the **CLI** and using **Openshift/Kubernetes objects (CRs)**, like *HostedCluster* and *NodePool* which are the top level objects that we will focus on. From both ones the proper Controller for that resource will take care of the object generation under that CR.

Reviewer: ```suggestion
In order to deploy a HostedCluster you have two options, using the **CLI** or using **Openshift/Kubernetes objects (CRs)**, like *HostedCluster* and *NodePool* which are the top-level API objects that we will focus on. From each, the proper controller will take care of the object generation under that CR.
```

This is how looks like when you have the Hypershift Operator deployed into the Openshift Management Cluster

```mermaid
flowchart LR
    style HO fill:#BAE0E9,stroke:#333,stroke-width:3px
    style HNDEP fill:#BAE0E9,stroke:#333,stroke-width:3px

    subgraph "Openshift Management Cluster"
        subgraph HN[Hypershift Namespace]
        HO[fa:fa-server  Hypershift-Operator] --> HNDEP[Deployments];
            subgraph HNDEP[Deployments]
            direction LR
                External-DNS --> descExD{{Manages Provider DNS *optional* }};

>                 Operator --> descOp{{Manages Hosted Clusters}};

Reviewer: What about NodePools?

Author: They come later :)

            end
        end
    end
```
> Legend:
>
> - Cyan: Provider agnostic OCP objects.

The CLI will create 2 deployments:

> - **ExternalDNS**: It will take care of the DNS management with the Cloud Provider. It will try to create entries into the domain/subdomain set in the install command.

Reviewer: How will it take care? can we describe or provide a reference?

- **Operator**: Stands for Hypershift Operator and it has some controllers inside which manages the HostedClusters and NodePools created in the cluster's namespace.

> > **NOTE:** Even using External DNS feature or not, as we said before, we will need to have an already registered DNS domain (or at least delegated) in *Route 53*. There, the Public entries will be created to allow access from outside.

Reviewer: ```suggestion
> **NOTE:** Regardless of whether you use the External DNS feature or not, as mentioned earlier, you need an already registered (or at least delegated) DNS domain in *Route 53*. This is where the public records are created to allow external access.
```

### HostedCluster

> This is the Hosted Cluster namespace where you as a user will create the definition of your cluster and your nodes. An easy way to create a hosted cluster it's just using the CLI as we've seen in the last sample.

Reviewer: ```suggestion
This is the Hosted Cluster namespace where you as a user will create the definition of your cluster and your nodes. An easy way to create a hosted cluster is by using the CLI as we've seen in the last sample.
```

> HostedCluster and NodePool controllers (part of Hypershift Operator) are monitoring the HostedCluster and NodePool objects present o created in the whole cluster. So once you create one of these components, the controllers will begin the reconciliation.

Reviewer: ```suggestion
HostedCluster and NodePool controllers (part of Hypershift Operator) are monitoring the HostedCluster and NodePool objects being created in the whole cluster. So once you create one of these components, the controllers will begin the reconciliation.
```

> You will need to have in mind that some more objects like Secrets and ConfigMaps (described in the diagram) are needed to achieve a succesfully HostedCluster deployment.

Reviewer: ```suggestion
You will need to have in mind that some more objects like Secrets and ConfigMaps (described in the diagram) are needed for a successful `HostedCluster` deployment.
```

From these mentioned objects, the Hypershift operator will create some more in 2 places, Openshift infrastructure and also in AWS Cloud Provider. Let's take a look to the following diagram to discover which objects are created in Openshift.

```mermaid
flowchart TB
    style HO fill:#BAE0E9,stroke:#333,stroke-width:3px
    style SC2 fill:#BAE0E9,stroke:#333,stroke-width:3px
    style CMC2 fill:#BAE0E9,stroke:#333,stroke-width:3px
    style NPO fill:#D3FFCE,stroke:#333,stroke-width:3px
    style HCO fill:#D3FFCE,stroke:#333,stroke-width:3px

>     AD[fa:fa-user Administrator]

Reviewer: Let's use the personas: https://hypershift-docs.netlify.app/reference/concepts-and-personas/ for consistency

            direction TB
            PRTN[Route]
            PVRIG[Route]
        end

        subgraph PBRT[Public Route Table]
            direction TB
            PBRIG[Route]
        end

        subgraph DNS
            direction TB
            PVZNL[Private Zone: .local]
            PVZNPD[Private Zone: public domain]
            PBZN[Public Zone]
        end

        subgraph IAM
            direction TB
            IAMR[Roles]
            IAMRP[Role Policies]
        end

        subgraph ELB[Elastic Load Balancer]
            direction TB
            IELB[Ingress ELB]
        end
    end
```

> Legend:
>
> - Orange: AWS Components
> - Cyan: Hypershift Operator
> - Red: Openshift Hosted Cluster

> **NOTE:** Just to remark somthing already mentioned, the Hypershift operator expects that the components in AWS exists. The only htings that are automatically created are the **Private Links** and the **Elastic Load Balancer**

### NodePools

The NodePool object creation implies a creation of a set of Openshift objects which triggers the NodePool controller actions.

Let's reuse the last diagram to dissect the object creation inside of Openshift:

> ```mermaid

Author: Nor HostedCluster or NodePool, this diagram it's only focused in the Hosted Control Plane NameSpace. Maybe I can reference the HC and NP. WDYT?

            end

            subgraph CMHCP[ConfigMaps]
                Configurations
                Certificates
                ...
            end

            subgraph SCS[Secrets]
                KBC[Kubeconfig]
                CA[Certificates]
                SCETC[...]
            end

            subgraph PVC[PVCs]
                etcd
                ovn
            end

            subgraph APPS[ControlPlane Apps]
                STS[Stateful Sets]
                DP[Deployments]
                RS[Replica Sets]
            end

            HCP <-.-> CL <-.-> AWSCL
            subgraph HCP[Hosted Control Plane]
                DescHCP("Contains the OCP
                HostedCluster details")
            end
            subgraph AWSCL[AWSCluster]
                DescAWSC("Contains the AWS details like
                VPC, Region, Subnet, etc...")
            end
            subgraph CL[Cluster]
                DescCL("Perform the relationship between
                AWSCluster and HostedControlPlane")
            end
        end
    end
```
> Legend:
>
> - Orange: OCP objects directly related with AWS.
> - Cyan: Provider agnostic OCP objects.

> At the same time, the Hypershift operator will expect to exists the mentioned Infra components in the defined provider (You could use the Hypershift CLI in order to create them automatically). As a first step it will assign an InfraID to all the objects created in the AWS side.

Reviewer 2: I'm having a hard time to follow this diagram. I'd expect us to present and list the cloud infra categorised in the following 4 buckets:
Pre-required unmanaged guest cluster aws account infra: e.g vpc (hypershift CLI is just a convenient automation for it, nothing else).
Infra managed by hypershift in management aws account . e.g LB for kas.
Infra managed by hypershift in guest cluster aws account. e.g nodepool instances.
Infra managed by kubernetes in guest cluster aws account. e.g PVs, service type LB.

Reviewer 2: For the Resource / controllers relationships we also have https://hypershift-docs.netlify.app/reference/controller-architecture/ already

Author: I've modified a bit the style and added a new section to show this 4 buckets you mention. PTAL and I'll make the corrections needed (if make sense) :)

Author: Also added a new plugin for MKDocs which enlarge images smoothly. Mainly for the Networking diagram gif

---
title: AWS Architecture and Resources
---

# AWS Architecture and Resources

> In this section we will talk about the AWS Resources and Openshift objects generated by Hypershift and how the Controllers interact with them during a Hosted Cluster deployment.

Reviewer 3: As I'm reading through this, I see several references to HyperShift. I believe the official name is hosted control planes. Is there a guideline about when "HyperShift" is correct and when "hosted control planes" is correct?

Author: Hey @lahinson Thank you for the comment, we decided to move this `index.md` file to other PR in order to push the priority in the HOSTEDCP-325. I will reference this comment in the new PR in order to continue with he conversation.
